### ISSUE TITLE
Mask propagation in TimeDistributed doesn't seem to work after Embedding

### ISSUE THREAD

**Issue by pdasigi** [fchollet/keras#3651]
```
from keras.layers import Input, Embedding, TimeDistributed

input_layer1 = Input(shape=(2,), dtype='int32')
embedding = Embedding(input_dim=10, output_dim=20, mask_zero=True)
output1 = embedding(input_layer1)
print embedding.get_output_mask_at(0)

input_layer2 = Input(shape=(2,5), dtype='int32')
td_embedding = TimeDistributed(embedding)
output2 = td_embedding(input_layer2)
print td_embedding.get_output_mask_at(0)
```

I am using `TimeDistributed(Embedding)` as shown above only to operate on higher order inputs, and propagate the shape information accordingly. However, when I run the code above, I see that the first print statement shows a (symbolic) mask (`Elemwise{neq,no_inplace}.0`) and the second one just gives a `None`. I'm not sure why `TimeDistributed` does not propagate the mask.
- [x] Check that you are up-to-date with the master branch of Keras. You can update with:
  pip install git+git://github.com/fchollet/keras.git --upgrade --no-deps
- [x] If running on Theano, check that you are up-to-date with the master branch of Theano. You can update with:
  pip install git+git://github.com/Theano/Theano.git --upgrade --no-deps
- [x] Provide a link to a GitHub Gist of a Python script that can reproduce your issue (or just copy the script here if it is short).

**Comment by fchollet** [comment]
What exactly did you expect to achieve?

Masking is for timeseries. The output of `Embedding` is a timeseries. If you apply `TimeDistributed` to `Embedding`, you get a timeseries of timeseries. What does it mean to mask such an output, according to you? Do you want to mask timesteps of the output, i.e. mask entire timeseries from the second level?

**Comment by pdasigi** [comment]
I am trying to embed a set of sentences. So the input to the embedding layer is of shape `(batch_size, num_sentences, num_words)` and I expect the output to be `(batch_size, num_sentences, num_words, embedding_dim)`. So yes, the output is indeed a timeseries of timeseries (with the two time dimensions being sentences and words). Would you recommend doing something else to achieve this?

I expected the output mask to be of shape `(batch_size, num_sentences, num_words)` as well, where each slice of the tensor (`mask[:, i, :]`) is the mask of the corresponding sentence. So I guess the units being masked are still the same (words), but the mask has an additional dimension.

**Comment by pdasigi** [comment]
Alternatively, changing the `get_output_shape_for` function in `Embedding` to return `input_shape + (output_dim,)` would do this. Is there a reason why `Embedding` assumes the input is 2D?

**Comment by xuewei4d** [comment]
Same here.
